### PR TITLE
Support for Node.js 0.12+

### LINKED ISSUE #65: Pipe the response as stream through gunzip in case of the compressed response

### LINKED ISSUE #80: Fix AdvancedAgent for Node.js 0.11
Non-persistent AdvancedAgent are behave as persistent when running on Node.js >0.11.6.

### LINKED ISSUE #93: Default status filter confuse new users
First time users doesn't understand why they receive `UNEXPECTED_STATUS_CODE` error, how to fix it and why they doesn't receive response body.

As discussed in the #91, improving documentation is not a perfect solution, so we must break the backward compatibility to save the time of newcomers.

### LINKED ISSUE #106: Do not enforce 1024 pool size on require
Require of `asker` have side-effects:

```js
var http = require('http');
console.log(http.Agent.defaultMaxSockets);
require('asker');
console.log(http.Agent.defaultMaxSockets);
```

Outputs:

```
5
1024
```

I believe [this code](https://github.com/nodules/asker/blob/master/lib/asker.js#L239-L245) is causing it.

### PR DESCRIPTION
**Pull request by narqo**
Major changes:

- move custom agent to [asker-advanced-agent] npm module

It feels like there is no need in this agent implementation in Node.js>=0.12 unless one wants to have `maxSocket`'s per agent class (e.g. marked with `name`) tuning, as well as agent creation caching. The solution itself really needs to be proofed as useful in Node.js>0.10. But will leave now it for future investigations.

- remove `onretry` option handler
- drop `isRetryAllowed` field from `statusFilter` option

The logic for `statusFilter` has been rethought as was suggested in #93. Internal `statusFilter` method has been renamed to `isNetworkError(code: Number) -> Boolean` and now by default returned `false` for status codes > 499, which always led to request retry-ment (if `options.maxRetries>0`).
All statuses greater than 200 are regarded as legit.
There is still no response redirection handling out the box, but it's usefulness is still under discussion.

/review @kaero 

- fixes #106 
- closes #93
- fixes #80
- fixes #65

[asker-advanced-agent]: https://github.com/nodules/asker-advanced-agent

### CODE REVIEW

File: lib/asker.js [hunk 1]
```diff
@@ -648,9 +580,20 @@ Request.prototype.execute = function(callback) {
...
     // start tracing total request execution time
     // including networks ops time, asker code execution and queue in the pool
-    contimer.start(this, this.buildTimerId('execution'));
+    contimer.start(this._timerCtx, this.buildTimerId('execution'));
 
-    this._tryHttpRequest(extend({}, this.options, { agent : Request.getAgent(this) }));
+    var agent = this.options.agent;
+
+    if (typeof agent === 'object' && typeof agent.addRequest !== 'function') {
+        agent = new AdvancedAgent(assign({ protocol : this.options.protocol }, agent));
```
Comment (narqo) [inline lib/asker.js]: todo: test agent creation

**Comment by Flackus** [comment]
I do like the changes but I should remind that this can't be shipped without updating the docs ;)
And migration guide (just a paragraph or two) would be splendid :)

**Comment by Flackus** [comment]
And last but not least, we should implement and stress-test it beforehand.
I think Realty would be a proper choice.

**Comment by Flackus** [comment]
I've just spoken with @kaero.
Some questions still need to be clarified. I'll arrange a meeting for a final decision.

File: lib/asker.js [hunk 2]
```diff
@@ -398,52 +358,45 @@ Request.prototype.done = function(err, data) {
...
-        headers : headers,
-        meta : meta ? extend(this.getResponseMetaBase(), meta) : this.getResponseMetaBase()
-    });
+Request.prototype.isNetworkError = function(code) {
+    if (typeof this._statusCodeFilter === 'function') {
+        var result = this._statusCodeFilter(code);
+        // @note invert `accept` result for backward compatibility with asker 0.6.x
+        return typeof result === 'object' ? (! result.accept) : result;
+    }
+    return code > 499 || code < 200;
```
Comment (Flackus) [inline lib/asker.js]: `code < 200` seems ti be unnecessary

**Comment by Flackus** [comment]
We've spoken today with @kaero and @asterx. Our points are:

1). Remove enforcing of `(global)Agent.maxSockets`
No questions whatsoever. Should be removed.

2). Remove `onretry` option handler
No questions either.

3). Redo `statusFilter` into two different options:
`isNetworkError`
`isRetryAllowed`
Both should return `Boolean`.

Changes in default `isNetworkError` in comparison with default `statusFilter.accept` should be described in the migration guide. Also, `code < 200` seems to be unnecessary.

We've also discussed backwards compatibility. Conclusion is to remove legacy `statusFilter` support :)

4). Keep advancedAgent in main module.
Arguments being:
* problem with host-port limit is still here
* `defaultAgent` is used by default, so if user doesn't care he has no worries
* this PR looses diff of `advanced_agent.js`, which is not cool at all.

So, if purpose of moving `advancedAgent` to separate module was to support <0.12 by emitting `socketRemoved`, that's one thing.
(we should still check, if it's not necessary now)

And if purpose was just to remove low-level tuning, we do not want that. This functionality should stay.

5). Doc changes and migration guide.
It just must happen :)

File: lib/asker.js [hunk 3]
```diff
@@ -176,7 +197,6 @@ Request.DEFAULT_OPTIONS = {
     headers : undefined,
     query : undefined,
     body : undefined,
-    onretry : undefined,
     statusFilter : undefined,
```
Comment (Flackus) [inline lib/asker.js]: new handlers should be added here

**Comment by narqo** [comment]
> So, if purpose of moving advancedAgent to separate module was to support <0.12 by emitting socketRemoved, that's one thing.
(we should still check, if it's not necessary now)
> And if purpose was just to remove low-level tuning, we do not want that. This functionality should stay.

The main purpose of moving advanced agent to separate module was that there is nothing asker specific in it (see [adavanced-agent tests](https://github.com/nodules/asker-advanced-agent/blob/master/test/test_pool.js)). 

Our original idea was to use [got](https://www.npmjs.com/package/got) as lowlevel http request module with asker's agent improvements. But it turned out that `got` is too highlevel for this purpose (it's impossible to add request/network timers as well as retries without monkey-patching got's internals). 

At the same time, advanced agent can be used with other "http-requesters" (e.g. `http.request`, `request` or`got`), because why not ;)

In general I don't understand your points about leaving it inside asker's codebase, as all requirements from (4) (except this diff problem, of course) are satisfied. Can't we just leave the diff problem out, the same way we do for custom unzip and retries implementations?

**Comment by narqo** [comment]
Two new options have been added:

1\. `isNetworkError` - checks whether passed status code should be treated as network error.

~~~js
isNetworkError(code Number) -> Bolean
~~~

2\. `isRetryAllowed` - checks if retry is allowed for passed error.

~~~js
isRetriesAllowed(error Error|AskerError) -> Boolean
~~~

`isRetriesAllowed` which takes `error` object is much more flexible, than the filter which takes statusCode.

See tests from `status_codes_processing` and `retries` for examples of reaching previous behaviour.

**Comment by Flackus** [comment]
> In general I don't understand your points about leaving it inside asker's codebase...

I'll wait for @kaero to reply.
But I'm still against this rationale, I don't see why we should extra-patch projects' codebase for the sake of such an abstraction.

File: lib/asker.js [hunk 4]
```diff
@@ -315,7 +288,7 @@ Request.prototype.formatTimestamp = function() {
     var timers = this.getTimers();
 
     // @todo let user to use custom format
-    return ['in ', timers.network || '0', '~', timers.total, ' ms'].join('');
+    return 'in ' + (timers.network || '0') + '~' + timers.total + ' ms';
```
Comment (Flackus) [inline lib/asker.js]: It can break some metrics, I think.
I vote for keeping previous format, but in addition you can do user-customisable format.

Comment (narqo) [inline lib/asker.js]: But the format haven't been changed here! It's about switching from `join('')` to string concatenation.

Comment (Flackus) [inline lib/asker.js]: Sorry, messed up a bit, caput quadratum :(

**Comment by Flackus** [comment]
@kaero ping :)

**Comment by kaero** [comment]
Move the advanced agent out of the asker. An agent setting interface kept, as i can see, so lgtm.

**Comment by Flackus** [comment]
Well, democracy in action.
@narqo should we release 1.0 Release Candidate? It should be stress-tested while in RC.

**Comment by narqo** [comment]
> should we release 1.0 Release Candidate. It should be stress-tested while in RC.

Sure. I'll make all the tests during this week.

**Comment by kaero** [comment]
we-e-e-ek :smile_cat:

**Comment by narqo** [comment]
@Flackus please check changes I've added to [readme] and new [migration] guide. Feel free to fix my grammar and typos ;)

[readme]: https://github.com/nodules/asker/blob/nodejs-next/README.md
[migration]: https://github.com/nodules/asker/blob/nodejs-next/MIGRATION.md

**Comment by 1999** [comment]
Could you describe 3 and 4 items of migration guide? Neither of them is clear to me.

File: MIGRATION.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+Version 1.0.0 introduced the following breaking changes:
+
+1\. `Agent.maxSockets` doesn't set to 1024 anymore.
+
+Since Node.js 0.12 the default value of `Agent.maxSockets` as well as `globalAgent.maxSockets` was set
+to `Infinity`. An enforcing of setting it to 1024 from the asker's code might lead to unwelcome behaviour
+for users which use Node.js 0.12 and above.
+
+Users who use Node.js 0.10 and don't use per server agent tuning should set `globalAgent.maxSockets` to
+1024 (or any acceptable value) by themself.
```
Comment (Flackus) [inline MIGRATION.md]: themselves, раз выше users. и без предлога by, он лишний

File: MIGRATION.md [hunk 6]
```diff
@@ -0,0 +1,65 @@
+# Migration
+
+## 1.0.0
+
+Version 1.0.0 introduced the following breaking changes:
+
+1\. `Agent.maxSockets` doesn't set to 1024 anymore.
```
Comment (1999) [inline MIGRATION.md]: isn't скорее

File: MIGRATION.md [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+Since Node.js 0.12 the default value of `Agent.maxSockets` as well as `globalAgent.maxSockets` was set
+to `Infinity`. An enforcing of setting it to 1024 from the asker's code might lead to unwelcome behaviour
+for users which use Node.js 0.12 and above.
+
+Users who use Node.js 0.10 and don't use per server agent tuning should set `globalAgent.maxSockets` to
```
Comment (1999) [inline MIGRATION.md]: Those who use, иначе масло масляное

File: MIGRATION.md [hunk 8]
```diff
@@ -0,0 +1,65 @@
+# Migration
+
+## 1.0.0
+
+Version 1.0.0 introduced the following breaking changes:
+
+1\. `Agent.maxSockets` doesn't set to 1024 anymore.
+
+Since Node.js 0.12 the default value of `Agent.maxSockets` as well as `globalAgent.maxSockets` was set
+to `Infinity`. An enforcing of setting it to 1024 from the asker's code might lead to unwelcome behaviour
```
Comment (Flackus) [inline MIGRATION.md]: Enforcing it within Asker could lead to an undesirable effect ... who use ... so it was removed (не хватает вывода)

File: MIGRATION.md (hunk 6 above)
Comment (Flackus) [inline MIGRATION.md]: Я бы еще добавил is not set, полная форма подчеркивает отрицание

File: MIGRATION.md (hunk 7 above)
Comment (Flackus) [inline MIGRATION.md]: :+1:

File: MIGRATION.md [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+for users which use Node.js 0.12 and above.
+
+Users who use Node.js 0.10 and don't use per server agent tuning should set `globalAgent.maxSockets` to
+1024 (or any acceptable value) by themself.
+
+See #106 for discussion.
+
+2\. `statusFilter` options has been split into two separate functions:
```
Comment (Flackus) [inline MIGRATION.md]: почему options? там была одна опция и одна функция. Так что множественное число лишнее

File: MIGRATION.md [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+- `isNetworkError(Number statusCode) -> Boolean`;
+- `isRetryAllowed(AskerError retryReason) -> Boolean`.
+
+The the code that uses asker<1.0.0
```
Comment (Flackus) [inline MIGRATION.md]: Да просто `asker < 1.0`

File: MIGRATION.md [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    return {
+        accept : ~[200, 201, 304].indexOf(code),
+        isRetryAllowed : code < 400 || code > 499
+    }
+}
+
+ask({ host: 'yandex.com', statusFilter: statusFilter });
+```
+
+the following changes have to be made on the project level:
```
Comment (Flackus) [inline MIGRATION.md]: Code above should be changed to:
Мне кажется, так понятнее.

File: MIGRATION.md [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+        return statusCode < 400 || statusCode > 499;
+    }
+}
+
+ask({ host: 'yandex.com', isNetworkError: isNetworkError, isRetryAllowed: isRetryAllowed });
+```
+
+See #93 for discussion of previous behaviour.
+
+3\. `onretry` event handler has been removed from the list of allowed options.
```
Comment (Flackus) [inline MIGRATION.md]: had been removed.
>Past Perfect Passive 
>Обозначает действие, которое совершилось (было совершено) к какому-то моменту в прошлом 

откуда -- не так важно.
создается впечатление, как будто внутри он остался и его кишки можно достать.

File: MIGRATION.md [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+
+4\. `agent.persistent` option has been removed.
```
Comment (Flackus) [inline MIGRATION.md]: тут тоже

File: MIGRATION.md (hunk 12 above)
Comment (Flackus) [inline MIGRATION.md]: отбой, я не прав. тут важен сам факт совершения, без указания момента
потерял практику :(

File: MIGRATION.md [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+
+---
+
+See CHANGELOG for the full list of changes introduced in 1.0.0.
```
Comment (Flackus) [inline MIGRATION.md]: я бы тут поспорил за артикль у changelog, но тут используется как имя файла, так что норм по идее

File: README.md [hunk 15]
```diff
@@ -4,14 +4,16 @@ Asker [![NPM version][npm-image]][npm-link] [![Build status][build-image]][build
 [![Dependency status][deps-image]][deps-link]
 [![devDependency status][devdeps-image]][devdeps-link]
 
-Asker is a wrapper for `http.request` method, which incorporates:
+Asker is a wrapper for `http.request`, which incorporates:
```
Comment (Flackus) [inline README.md]: :+1:

File: README.md [hunk 16]
```diff
@@ -4,14 +4,16 @@ Asker [![NPM version][npm-image]][npm-link] [![Build status][build-image]][build
...
+
 * response deflating using gzip,
 * requests retrying,
 * connection pools tuning.
 
-If you are looking for a module to fetch 3rd-party web content (pages, RSS, files or something else), don't waste your time and look at the [request](http://npm.im/request) module, because `asker` doesn't support cookies and redirects out of the box.
+If you are looking for a module to fetch 3rd-party web content (pages, RSS, files or something else), don't waste your
+time and look at the [request](http://npm.im/request) module, as asker doesn't support cookies and redirects out of the box.
```
Comment (Flackus) [inline README.md]: кавычки у asker я бы оставил

File: README.md [hunk 17]
```diff
@@ -141,18 +171,18 @@ ask({
 });
 ```
 
-If you need to send multiple files in single field wrap them in an array.
+You can send multiple files in a single field if wrap them in an array:
```
Comment (Flackus) [inline README.md]: if wrap them как-то по-русски очень.
сходу, правда, лучше не придумал. утро вечера мудренее.

File: README.md [hunk 18]
```diff
@@ -207,37 +242,62 @@ Asker.bodyEncoders.trimText = function(body, setContentType) {
 
 ### The problem
 
-At this point of time (version `0.10` and below) node.js provides a socket pool manager that works as follows:
+For Node.js versions 0.10 and below an internal socket pool manager works as follows:
```
Comment (Flackus) [inline README.md]: стоп. мы же вроде выясняли, что он так до сих пор и работает, разве нет?
ну и раньше было все же чуть понятнее, я бы оставил provides ... that

File: README.md [hunk 19]
```diff
@@ -207,37 +242,62 @@ Asker.bodyEncoders.trimText = function(body, setContentType) {
...
 
 * by default `globalAgent` is used for all outgoing http requests;
 * each `Agent` instance, including `globalAgent`, has a `maxSockets` property, which you can change;
-* socket limit is set for each unique __host-port pair__, that is served by this particular `Agent`.
+* socket limit is set for each unique **host:port pair**, that is served by this particular `Agent`.
 
-That is sometimes an unwanted behaviour. Let's take an example.
+The pool may behaves unexpectedly in some cases. For example, you have two backend servers: `backend:3000` and `backend:4000`.
```
Comment (Flackus) [inline README.md]: behave

File: README.md [hunk 20]
```diff
@@ -207,37 +242,62 @@ Asker.bodyEncoders.trimText = function(body, setContentType) {
...
+First backend is indispensable for the application, but the second one is complementary (e.g. it makes an http call for advertisements you show later).
 
-You have two backends: `backend:3000` and `backend:4000`. First backend is indispensable for the application, but the second one is complementary. E.g., it makes http calls for the advertisements that you show later.
+Under the heavy load this additional backend (which is usually less fault-tolerant) may occupy all sockets that OS provides
+for the whole Node.js process, because default pool manager cares only about **host:port pairs** and `defaultAgent`
+does not correct each backend's socket limit according to process limit.
 
-Under heavy load this additional backend (which is usually less fault-tolerant) may occupy all sockets that OS provides for the whole `node` process, because default pool manager cares only about __host-port__ pairs. `defaultAgent` does not anyhow correct each backend' socket limit according to process limit.
+In the scenario above both backends will be unavailable for us, even if first indispensable server works fine.
```
Comment (Flackus) [inline README.md]: for us я бы убрал.
if the first ... backend (вместо server)

File: README.md [hunk 21]
```diff
@@ -207,37 +242,62 @@ Asker.bodyEncoders.trimText = function(body, setContentType) {
...
 
 ### Solution
 
-How `Asker` can help you manage this problem? `Asker` can create a custom instance of `http.Agent` for any given backend. And you can set up a `maxSockets` property by calculating each backend priority.
+How asker can help you manage this problem? You can create a custom instance of `http.Agent` for any given backend.
+And you can set a `maxSockets` property by calculating each backend priority. Asker gives you an API to help with that.
```
Comment (Flackus) [inline README.md]: https://www.google.ru/?gfe_rd=cr&ei=QjRBVuG9MpTANOz8segO&gws_rd=ssl#newwindow=1&safe=off&q=gives+you+API

File: README.md [hunk 22]
```diff
@@ -207,37 +242,62 @@ Asker.bodyEncoders.trimText = function(body, setContentType) {
...
 
 ### API
 
 `{Object} agent` `http.Agent` options:
 
-* `{String} name='globalAgent'` unique name for this backend
-* `{Number} maxSockets=1024` socket limit for this backend
-* `{Boolean} persistent=true` either agent would be deleted when it's last socket is removed. You can consider setting it to false, if you create agent's `name` in runtime.
+* `{String} name='globalAgent'` – Unique name for the asked server.
+* `{Number} maxSockets` – Socket limit for the server. For Node.js versions previous to 0.12 default value is `5`.
```
Comment (Flackus) [inline README.md]: `for node.js 0.12 and below`?

File: README.md [hunk 23]
```diff
@@ -207,37 +242,62 @@ Asker.bodyEncoders.trimText = function(body, setContentType) {
...
+
+#### Example
```
Comment (Flackus) [inline README.md]: :+1:

**Comment by Flackus** [comment]
@narqo great job on the transltation :+1:

File: README.md (hunk 16 above)
Comment (narqo) [inline README.md]: Там по тексту встречались: asker, `asker`, Asker. Первое выглядит лучше и соотносится с «request» и «Terror», которые в тексте без тиков (`). С маленькой буквы, потому что nodejs-way.

File: README.md (hunk 17 above)
Comment (narqo) [inline README.md]: > Wrap the fields in an array if you want to send multiple files.

?

> To send multiple files wrap the fields in an array as shown in the example below.

?

File: README.md (hunk 22 above)
Comment (narqo) [inline README.md]: в 0.12 — `Infinity`, поправил на 

> Node.js 0.10 and below sets it to `5` by default.

**Comment by narqo** [comment]
> Could you describe 3 and 4 items of migration guide? Neither of them is clear to me.

@1999 I can't think of any additional description except that this options were removed and are no longer acceptable.

File: README.md (hunk 16 above)
Comment (Flackus) [inline README.md]: Принято. Везед поправил?

File: README.md (hunk 17 above)
Comment (Flackus) [inline README.md]: Первый вариант вроде ок.

**Comment by narqo** [comment]
done!

**Comment by Flackus** [comment]
Ship it!

**Comment by narqo** [comment]
So, let it be!

**Pull request merged**